File: articles/norbert-wiener-cybernetics-and-feedback-in-computation/haskell/NegativeFeedback.hs
module Main where

step :: Double -> Double -> Double -> Double
step target gain x = x - gain * (x - target)

run :: Int -> Double -> Double -> Double -> Double
run 0 _ _ x = x
run n target gain x = run (n - 1) target gain (step target gain x)

main :: IO ()
main = putStrLn ("final_state=" ++ show (run 5 0.0 0.2 10.0))

File: articles/norbert-wiener-cybernetics-and-feedback-in-computation/c/negative_feedback.c
#include <stdio.h>

int main(void) {
    double x = 10.0;
    double target = 0.0;
    double gain = 0.2;
    for (int i = 0; i < 5; i++) {
        x = x - gain * (x - target);
    }
    printf("final_state=%.6f\n", x);
    return 0;
}
